tests/test_upload.py

```diff
@@ -57,6 +59,104 @@ def test_successful_upload(make_settings, capsys):
     assert captured.out.count(NEW_RELEASE_URL) == 1
 
 
+def test_successful_upload_add_gpg_signature(make_settings, capsys, monkeypatch):
+    """Test uploading package when gps signature are added to it"""
+
+    replaced_add_gpg_signature = pretend.call_recorder(
+        lambda signature_filepath, signature_filename: None
+    )
+
+    package = package_file.PackageFile(
+        filename=helpers.WHEEL_FIXTURE,
+        comment=None,
+        metadata=pretend.stub(name="twine"),
+        python_version=None,
+        filetype=None,
+    )
+    package.signed_basefilename = "twine.asc"
+
+    monkeypatch.setattr(package, "add_gpg_signature", replaced_add_gpg_signature)
```

Comment:
We're monkeypatching so much I don't even know what we're supposed to be testing based on the test description

Reply:
I am testing the branch of `twine.upload.upload` which adds gpg signatures to package

https://github.com/pypa/twine/blob/ae2b876fa20813d979c01fbb341cd12b1d74a47c/twine/commands/upload.py#L82-L84

To that effect, I am using a custom package attribute and created a call recorder to `package.add_gpg_signature` to test calls to that branch, and patched a few functions in `upload` class to use our custom package class.

I have also added some comments in the code as well to help in understanding the code.

Reply:
Adding myself as a reviewer to this; I'm also wary of of the amount of patching going on (though I'm assuming you have good intentions).

Reply:
So the primary aim of all the  monkey-patching is to be able to test the particular branch of code mentioned above. I felt that the best way of testing is to patch up certain sections of [twine.commands.upload:upload](https://github.com/pypa/twine/blob/ae2b876fa20813d979c01fbb341cd12b1d74a47c/twine/commands/upload.py#L53) which help in this.

So that involves patching https://github.com/pypa/twine/blob/ae2b876fa20813d979c01fbb341cd12b1d74a47c/twine/commands/upload.py#L54

 to return an `.asc` file to act as a signature.

and patching https://github.com/pypa/twine/blob/ae2b876fa20813d979c01fbb341cd12b1d74a47c/twine/commands/upload.py#L68-L70

returns the package object created in the test with the `signed_basefilename`

Finally we patch https://github.com/pypa/twine/blob/ae2b876fa20813d979c01fbb341cd12b1d74a47c/twine/commands/upload.py#L84 to record the call made to it.

I hope that explains the idea behind the patching, and I have also added comments in the code to expand on that. Do let me know if the comments have to be expanded upon more for clarity.